Blog Post: From OpenAI to Open LLMs with Messages API on Hugging Face (#1804)
huggingface/blog: tgi-messages-api.md

---
title: "From OpenAI to Open LLMs with Messages API on Hugging Face" 
thumbnail: /blog/assets/tgi-messages-api/thumbnail.png
authors:
- user: philschmid
- user: andrewrreed
- user: Jofthomas
---

# From OpenAI to Open LLMs with Messages API on Hugging Face

> ## Introduction

Reviewer: ```suggestion
```

> *"The new Messages API with OpenAI compatibility makes it easy for Ryght's real-time GenAI orchestration platform to switch LLM use cases from OpenAI to open models. Our migration from GPT4 to Mixtral/Llama2 on Inference Endpoints is effortless, and now we have a simplified workflow with more control over our AI solutions." - [Johnny Crupi, Ryght CTO](https://www.linkedin.com/in/johncrupi/)*

The new Messages API is now available in Inference Endpoints (dedicated and serverless). To get you started quickly, we’ve included detailed examples of how to:

- Create an Inference Endpoint
- Query it with OpenAI’s SDK
- Integrate with LangChain and LlamaIndex

***Limitations:** The Messages API does not currently support function calling and will only work for LLMs with a `chat_template` defined in their [tokenizer](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1/blob/main/tokenizer_config.json).*

## **Create an Inference Endpoint using `huggingface_hub`**

[Inference Endpoints](https://huggingface.co/docs/inference-endpoints/index) offers a secure, production solution to easily deploy any Transformers model from the Hub on dedicated infrastructure managed by Hugging Face.

> Here, we'll use it to host the powerful [Nous-Hermes-2-Mixtral-8x7B-DPO](https://huggingface.co/NousResearch/Nous-Hermes-2-Mixtral-8x7B-DPO) as an endpoint running on [Text Generation Inference](https://huggingface.co/docs/text-generation-inference/index), our high performance inference solution for serving LLMs in production. 

Reviewer: ```suggestion
In this example, we will deploy [Nous-Hermes-2-Mixtral-8x7B-DPO](https://huggingface.co/NousResearch/Nous-Hermes-2-Mixtral-8x7B-DPO), a fine-tuned Mixtral model to Inference Endpoints using [Text Generation Inference](https://huggingface.co/docs/text-generation-inference/index).
```

    framework="pytorch",
    task="text-generation",
    accelerator="gpu",
    vendor="aws",
    region="us-east-1",
    type="protected",
    instance_type="p4de",
    instance_size="2xlarge",
    custom_image={
        "health_route": "/health",
        "env": {
            "MAX_INPUT_LENGTH": "4096",
            "MAX_BATCH_PREFILL_TOKENS": "4096",
            "MAX_TOTAL_TOKENS": "32000",
            "MAX_BATCH_TOTAL_TOKENS": "1024000",
            "MODEL_ID": "/repository",
        },
        # "url": "ghcr.io/huggingface/text-generation-inference:1.4.0",  # must be >= 1.4.0
        "url": "ghcr.io/huggingface/text-generation-inference:sha-ee1cf51",
    },
)

endpoint.wait()
print(endpoint.status)

```

It will take a few minutes for our deployment to spin up. We can utilize the `.wait()` utility to block the running thread until the endpoint reaches a final "running" state. Once running, we can confirm its status and take it for a spin via the UI Playground:

![IE UI Overview](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/messages-api/endpoint-overview.png)

Great, we now have a working deployment! 

> >💡 By default, your endpoint will scale-to-zero after 15 minutes of idle time without any requests to optimize cost during periods of inactivity. Check out [the Hub Python Library documentation](https://huggingface.co/docs/huggingface_hub/guides/inference_endpoints) to see all the functionality available for managing your endpoint lifecycle.

Reviewer: As said before i think thats not true

Reviewer: @Wauplin can you confirm? Through the UI its not the case.

Reviewer 2: @philschmid It is indeed the case (see `min_replicas=0` by default [here](https://github.com/huggingface/huggingface_hub/blob/6b202f198a94609d4a907fd26c65307d275a2d3d/src/huggingface_hub/hf_api.py#L6991)). Didn't know it wasn't the case by default through the UI. I can make a PR to update this in `huggingface_hub` but it would be a breaking change now :confused: (better to align now than in the future though)

Reviewer: yes lets make sure the comment says, when using `huggingface_hub`... endpoints will scale....

We can deploy the model in just [a few clicks from the UI](https://ui.endpoints.huggingface.co/new?vendor=aws&repository=NousResearch%2FNous-Hermes-2-Mixtral-8x7B-DPO&tgi_max_total_tokens=32000&tgi=true&tgi_max_input_length=1024&task=text-generation&instance_size=2xlarge&tgi_max_batch_prefill_tokens=2048&tgi_max_batch_total_tokens=1024000&no_suggested_compute=true&accelerator=gpu&region=us-east-1), or take advantage of the `huggingface_hub` Python library to programmatically create and manage Inference Endpoints. We demonstrate the use of the Hub library below.

In our API call shown below, we need to specify the endpoint name and model repository, along with the task of `text-generation`. In this example we use a `protected` type so access to the deployed endpoint will require a valid Hugging Face token. We also need to configure the hardware requirements like vendor, region, accelerator, instance type, and size. You can check out the list of available resource options [using this API call](https://api.endpoints.huggingface.cloud/#get-/v2/provider), and view recommended configurations for select models in our catalog [here](https://ui.endpoints.huggingface.co/catalog). 

> 

Reviewer: ```suggestion

_Note: You might need to request a quota upgrade via email to [api-enterprise@huggingface.co](mailto:api-enterprise@huggingface.co)_ 

```

_Note: You may need to request a quota upgrade by sending an email to [api-enterprise@huggingface.co](mailto:api-enterprise@huggingface.co)_ 

```python
from huggingface_hub import create_inference_endpoint

endpoint = create_inference_endpoint(
    "nous-hermes-2-mixtral-8x7b-demo",
    repository="NousResearch/Nous-Hermes-2-Mixtral-8x7B-DPO",
    framework="pytorch",
    task="text-generation",
    accelerator="gpu",
    vendor="aws",
    region="us-east-1",
    type="protected",
    instance_type="p4de",
    instance_size="2xlarge",
    custom_image={
        "health_route": "/health",
        "env": {
            "MAX_INPUT_LENGTH": "4096",
            "MAX_BATCH_PREFILL_TOKENS": "4096",
            "MAX_TOTAL_TOKENS": "32000",
            "MAX_BATCH_TOTAL_TOKENS": "1024000",
            "MODEL_ID": "/repository",
        },

>         "url": "ghcr.io/huggingface/text-generation-inference:sha-1734540",

Author: @philschmid Is it ok that we need to use a non tagged version here? I know its handled in the IE UI, but its very visible since our example deploys with the API...

Reviewer: Yes, we can add a comment maybe.

Author: https://github.com/huggingface/blog/pull/1804/commits/f099445394ad964c8a1c357105e7baf2ab065ee9

---
title: "From OpenAI to Open LLMs with Messages API on Hugging Face" 
thumbnail: /blog/assets/tgi-messages-api/thumbnail.png
authors:
- user: andrewrreed
- user: philschmid

> - user: Jofthomas

Reviewer: ```suggestion
- user: jofthomas
```

maybte that solves it?

Author: Didn't work. I'm not sure what database the users are being queried/matched against, but I think the problem is that database is not updates with people new to HF.

For example, on the [LangChain agents blog post](https://huggingface.co/blog/open-source-llms-as-agents), Aymeric and Joffrey appear in the production blog header, but also do not appear in the CodeSpaces version:

![image](https://github.com/huggingface/blog/assets/23462538/667e2bc1-e7d4-4ee3-9562-ca7adc9b4820)

For this reason, I think this problem will be corrected once we merge

Great, we now have a working endpoint! 

>💡 When deploying with `huggingface_hub`, your endpoint will scale-to-zero after 15 minutes of idle time by default to optimize cost during periods of inactivity. Check out [the Hub Python Library documentation](https://huggingface.co/docs/huggingface_hub/guides/inference_endpoints) to see all the functionality available for managing your endpoint lifecycle.

## Using Inference Endpoints with OpenAI client libraries

Messages support in TGI makes Inference Endpoints directly compatible with the OpenAI Chat Completion API. This means that any existing scripts that use OpenAI models via the OpenAI client libraries can be directly swapped out to use any open LLM running on a TGI endpoint!

With this seamless transition, you can immediately take advantage of the numerous benefits offered by open models:

- Complete control and transparency over models and data
- No more worrying about rate limits
- The ability to fully customize systems according to your specific needs

Lets see how.

### With the Python client

> The example below shows how to make this transition using the [OpenAI Python Library](https://github.com/openai/openai-python). Simply replace the `<ENDPOINT_URL>` with your endpoint URL (be sure to include the `v1/` the suffix) and populate the `<HF_API_TOKEN>` field with a valid Hugging Face user token. The `<ENDPOINT_URL>` can be gathered from Inference Endpoints UI, or from the endpoint object we created above with `endpoint.url`.

Reviewer 3: ```suggestion
The example below shows how to make this transition using the [OpenAI Python Library](https://github.com/openai/openai-python). Simply replace the `<ENDPOINT_URL>` with your endpoint URL (be sure to include the `v1/` suffix) and populate the `<HF_API_TOKEN>` field with a valid Hugging Face user token. The `<ENDPOINT_URL>` can be gathered from Inference Endpoints UI, or from the endpoint object we created above with `endpoint.url`.
```

We can then use the client as usual, passing a list of messages to stream responses from our Inference Endpoint.

```python
from openai import OpenAI

# initialize the client but point it to TGI
client = OpenAI(
    base_url="<ENDPOINT_URL>" + "/v1/",  # replace with your endpoint url
    api_key="<HF_API_TOKEN>",  # replace with your token
)
chat_completion = client.chat.completions.create(
    model="tgi",
    messages=[

> 		{"role": "system", "content": "You are a helpful assistant."},
>         {"role": "user", "content": "Why is open-source software important?"},

Reviewer 3: ```suggestion
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Why is open-source software important?"},
```

    ],

>     stream=True,
> 	max_tokens=500

Reviewer 3: ```suggestion
    stream=True,
     max_tokens=500
```

)
from llama_index import download_loader
from llama_index.embeddings import HuggingFaceEmbedding
from llama_index.query_engine import CitationQueryEngine

SimpleWebPageReader = download_loader("SimpleWebPageReader")

documents = SimpleWebPageReader(html_to_text=True).load_data(
    ["https://huggingface.co/blog/open-source-llms-as-agents"]
)

# Load embedding model
embed_model = HuggingFaceEmbedding(model_name="BAAI/bge-large-en-v1.5")

# Pass LLM to pipeline
service_context = ServiceContext.from_defaults(embed_model=embed_model, llm=llm)
index = VectorStoreIndex.from_documents(
    documents, service_context=service_context, show_progress=True
)

# Query the index
query_engine = CitationQueryEngine.from_args(
    index,
    similarity_top_k=2,
)
response = query_engine.query(
    "According to this article which open-source model is the best for an agent behaviour?"
)
```

```
According to the article, Mixtral-8x7B is the best performing open-source model for an agent behavior [5]. It even beats GPT-3.5 in this task. However, it's worth noting that Mixtral's performance could be further improved with proper fine-tuning for function calling and task planning skills [5].

> ```

Reviewer 3: What does `[5]` refer to? I'd just hyperlink "According to this article" to the source. Also, are the code backquotes intentional?

Author: This text blurb is the output from the `query_engine.query()` line above. The [5] is the citation is a reference to the text chunk in the vector store from which the generation was based on

Reviewer 3: Ah sorry, read it too fast!

---
title: "From OpenAI to Open LLMs with Messages API on Hugging Face" 
thumbnail: /blog/assets/tgi-messages-api/thumbnail.png
authors:
- user: philschmid
- user: andrewrreed
- user: Jofthomas
---

# From OpenAI to Open LLMs with Messages API on Hugging Face

## Introduction

> We are excited to introduce the Messages API to provide OpenAI compatibility with Text Generation Inference (TGI) and Inference Endpoints.

Reviewer 3: ```suggestion
We are excited to introduce the Messages API to provide OpenAI compatibility to Text Generation Inference (TGI) and Inference Endpoints.
```

Reviewer 3: "We provide compatibility..." sounds similar to me to "We add compatibility...", but I guess `with` works as well in this case.

Starting with version 1.4.0, TGI offers an API compatible with the OpenAI Chat Completion API. The new Messages API allows customers and users to transition seamlessly from OpenAI models to open LLMs. The API can be directly used with OpenAI's client libraries or third-party tools, like LangChain or LlamaIndex.

> *“Switching from Azure OpenAI to Hugging Face Inference Endpoints and Mixtral using the new Message API with OpenAI compatibility was a game-changer for us. The seamless transition and compatibility ensured performance and gave us more flexibility and control.” - RyghtAI*

Author: This is a placeholder quote - waiting on the real one from the customer.

> The new Messages API is now available in Inference Endpoints (dedicated and serverless). To get you started quickly, we’ve included detailed examples of how to:

Reviewer 3: ```suggestion
The new Messages API is also available in Inference Endpoints now, on both dedicated and serverless flavors. To get you started quickly, we’ve included detailed examples of how to:
```

- Create an Inference Endpoint

> - Use OpenAI’s SDK

Reviewer 3: ```suggestion
- Using Inference Endpoints with OpenAI client libraries
```

Also, maybe link to anchors in this ToC.

Author: Good call! https://github.com/huggingface/blog/pull/1804/commits/3a49f02257ec43ef3f28745018a5525b372e2216

- Integrate with LangChain and LlamaIndex

> ***Limitations:** The Messages API does not currently support function calling and will only work for LLMs with a `chat_template` defined in their [tokenizer](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1/blob/main/tokenizer_config.json).*

Reviewer 3: ```suggestion
**Limitations:** The Messages API does not currently support function calling and will only work for LLMs with a `chat_template` defined in their tokenizer configuration, like in the case of [Mixtral 8x7B Instruct](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1/blob/125c431e2ff41a156b9f9076f744d2f35dd6e67a/tokenizer_config.json#L42).
```

> ## **Create an Inference Endpoint using `huggingface_hub`**

Reviewer 3: ```suggestion
## Create an Inference Endpoint
```

[Inference Endpoints](https://huggingface.co/docs/inference-endpoints/index) offers a secure, production solution to easily deploy any Transformers model from the Hub on dedicated infrastructure managed by Hugging Face.

Here, we'll use it to host the powerful [Nous-Hermes-2-Mixtral-8x7B-DPO](https://huggingface.co/NousResearch/Nous-Hermes-2-Mixtral-8x7B-DPO) as an endpoint running on [Text Generation Inference](https://huggingface.co/docs/text-generation-inference/index), our high performance inference solution for serving LLMs in production. 

We can deploy the model in just [a few clicks from the UI](https://ui.endpoints.huggingface.co/new?vendor=aws&repository=NousResearch%2FNous-Hermes-2-Mixtral-8x7B-DPO&tgi_max_total_tokens=32000&tgi=true&tgi_max_input_length=1024&task=text-generation&instance_size=2xlarge&tgi_max_batch_prefill_tokens=2048&tgi_max_batch_total_tokens=1024000&no_suggested_compute=true&accelerator=gpu&region=us-east-1), or take advantage of the `huggingface_hub` Python library to programmatically create and manage Inference Endpoints. We demonstrate the use of the Hub library below.

> First, we need to specify the endpoint name and model repository, along with the task of text-generation. A protected Inference Endpoint means a valid HF token is required to access the deployed API. We also need to configure the hardware requirements like vendor, region, accelerator, instance type, and size. You can check out the list of available resource options [here](https://api.endpoints.huggingface.cloud/#get-/v2/provider) and view recommended configurations for select models in our catalog [here](https://ui.endpoints.huggingface.co/catalog). 

Reviewer 3: ```suggestion
In our API call shown below, we need to specify the endpoint name and model repository, along with the task of `text-generation`. In this example we use a `protected` type so access to the deployed endpoint will require a valid Hugging Face token. We also need to configure the hardware requirements like vendor, region, accelerator, instance type, and size. You can check out the list of available resource options [using this API call](https://api.endpoints.huggingface.cloud/#get-/v2/provider), and view recommended configurations for select models in our catalog [here](https://ui.endpoints.huggingface.co/catalog). 
```

```python
from huggingface_hub import create_inference_endpoint

endpoint = create_inference_endpoint(
    "nous-hermes-2-mixtral-8x7b-demo",
    repository="NousResearch/Nous-Hermes-2-Mixtral-8x7B-DPO",
    framework="pytorch",
    task="text-generation",
    accelerator="gpu",
    vendor="aws",
    region="us-east-1",
    type="protected",
    instance_type="p4de",
    instance_size="2xlarge",
    namespace="HF-test-lab",
    custom_image={
        "health_route": "/health",
        "env": {
            "MAX_INPUT_LENGTH": "4096",
            "MAX_BATCH_PREFILL_TOKENS": "4096",
            "MAX_TOTAL_TOKENS": "32000",
            "MAX_BATCH_TOTAL_TOKENS": "1024000",
            "MODEL_ID": "/repository",
        },
        # "url": "ghcr.io/huggingface/text-generation-inference:1.4.0",  # must be >= 1.4.0

>         "url": "ghcr.io/huggingface/text-generation-inference:sha-ee1cf51",

Author: We should hold off on publishing the blog until [this fix](https://github.com/huggingface/text-generation-inference/commit/ee1cf51ce796e4b034eedaf3e909b4c902eae70c) is included in a patch release (and Inference Endpoints is updated to reflect the new TGI container)

    },
)

endpoint.wait()
print(endpoint.status)

```

> It will take a few minutes for our deployment to spin up. We can utilize the `.wait()` utility to block the running thread until the endpoint reaches a final "running" state. Once running, we can confirm its status and take it for a spin via the UI Playground:

Reviewer 3: ```suggestion
It will take a few minutes for our deployment to spin up. We can use the `.wait()` utility to block the running thread until the endpoint reaches a final "running" state. Once running, we can confirm its status and take it for a spin via the UI Playground:
```

> > PLACEHOLDER FOR IMAGE

Author: Need link to image

Author: done

Great, we now have a working deployment! 

> <aside>

Reviewer 3: Not sure how this renders, to be checked.

Author: Yep fixed this with a block quote - its odd that thats how Notion copy/paste exports it 🤔

💡 By default, your endpoint will scale-to-zero after 15 minutes of idle time without any requests to optimize cost during periods of inactivity. Check out [the Hub Python Library documentation](https://huggingface.co/docs/huggingface_hub/guides/inference_endpoints) to see all the functionality available for managing your endpoint lifecycle.

</aside>

> ## Using Inference Endpoints via OpenAI client libraries

Reviewer 3: ```suggestion
## Using Inference Endpoints with OpenAI client libraries
```

> The added support for messages in TGI makes Inference Endpoints directly compatible with the OpenAI Chat Completion API. This means that any existing scripts that use OpenAI models via the OpenAI client libraries can be directly swapped out to use any open LLM running on a TGI endpoint!

Reviewer 3: ```suggestion
Messages support in TGI makes Inference Endpoints directly compatible with the OpenAI Chat Completion API. This means that any existing scripts that use OpenAI models via the OpenAI client libraries can be directly swapped out to use any open LLM running on a TGI endpoint!
```

With this seamless transition, you can immediately take advantage of the numerous benefits offered by open models:

- Complete control and transparency over models and data
- No more worrying about rate limits
- The ability to fully customize systems according to your specific needs

Lets see how.

### With the Python client

> The example below shows how to make this transition using the [OpenAI Python Library](https://github.com/openai/openai-python). Simply replace the `<ENDPOINT_URL>` with your endpoint URL (be sure to include the `v1/` the suffix) and populate the `<HF_API_TOKEN>` field with a valid Hugging Face user token.

Reviewer 3: ```suggestion
The example below shows how to make this transition using the [OpenAI Python Library](https://github.com/openai/openai-python). Simply replace the `<ENDPOINT_URL>` with your endpoint URL (be sure to include the `v1/` suffix) and populate the `<HF_API_TOKEN>` field with a valid Hugging Face user token.
```

Maybe explain where they can get the endpoint url. Is it displayed by `endpoint.status` above? Maybe also insist that the hf token is necessary because the endpoint is `protected`. Or is it always needed?

Author: Great suggestion. Added here: https://github.com/huggingface/blog/pull/1804/commits/3aceba1a2ff55d914134f58a40a287f9a540ad91

We can then use the client as usual, passing a list of messages to stream responses from our Inference Endpoint.

```python
from openai import OpenAI

# initialize the client but point it to TGI
client = OpenAI(
    base_url="<ENDPOINT_URL>" + "/v1/",  # replace with your endpoint url
    api_key="<HF_API_TOKEN>",  # replace with your token
)
chat_completion = client.chat.completions.create(
    model="tgi",
    messages=[

> 				{"role": "system", "content": "You are a helpful assistant."},
>         {"role": "user", "content": "Why is open-source software important?"},

Reviewer 3: ```suggestion
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Why is open-source software important?"},
```

    ],

>     stream=True,
> 		max_tokens=500

Reviewer 3: ```suggestion
    stream=True,
    max_tokens=500
```

)

# iterate and print stream
for message in chat_completion:
    print(message.choices[0].delta.content, end="")
```

> Behind the scenes, TGI’s Messages API automatically converts the list of messages into the model’s required instruction format using it’s [chat template](https://huggingface.co/docs/transformers/chat_templating). 

Reviewer 3: ```suggestion
Behind the scenes, TGI’s Messages API automatically converts the list of messages into the model’s required instruction format using its [chat template](https://huggingface.co/docs/transformers/chat_templating). 
```

<aside>
💡 Certain OpenAI features, like function calling, are not compatible with TGI. Currently, the Messages API supports the following chat completion parameters: `stream`, `max_new_tokens`, `frequency_penalty`, `logprobs`, `seed`, `temperature`, and `top_p`.

</aside>

> ### With the Javascript client

Reviewer 3: ```suggestion
### With the JavaScript client
```

Here’s the same streaming example above, but using the [OpenAI Javascript/Typescript Library](https://github.com/openai/openai-node).

```jsx
import OpenAI from "openai";

const openai = new OpenAI({
  baseURL: "<ENDPOINT_URL>" + "/v1/", // replace with your endpoint url
  apiKey: "<HF_API_TOKEN>", // replace with your token
});

async function main() {
  const stream = await openai.chat.completions.create({
    model: "tgi",
    messages: [
      { role: "system", content: "You are a helpful assistant." },
      { role: "user", content: "Why is open-source software important?" },
    ],
    stream: true,
    max_tokens: 500,
  });
  for await (const chunk of stream) {
    process.stdout.write(chunk.choices[0]?.delta?.content || "");
  }
}

main();
```

> ## How to use with LangChain

Reviewer 3: I'd maybe use a level-2 header for "How to integrate with LangChain and LlamaIndex" and then subsections for each.

Reviewer 3: Also, not sure why the section title is not displayed highlighted, maybe there's some unfinished string in the previous code blocks?

Author: Done: https://github.com/huggingface/blog/pull/1804/commits/803ba7bfc79ef1af607c5adc02df7c558b437c9b

> Now, let’s see how to use this newly created endpoint with your preferred RAG framework. To use it in [LangChain](https://python.langchain.com/docs/get_started/introduction), simply create an instance of `ChatOpenAI` and pass your `<ENDPOINT_URL>`  and  `<HF_API_TOKEN>`  as follows:

Reviewer 3: ```suggestion
Now, let’s see how to use this newly created endpoint with your preferred RAG framework. To use it in [LangChain](https://python.langchain.com/docs/get_started/introduction), simply create an instance of `ChatOpenAI` and pass your `<ENDPOINT_URL>` and `<HF_API_TOKEN>` as follows:
```

```python
from langchain_community.chat_models.openai import ChatOpenAI

llm = ChatOpenAI(
    model_name="tgi",
    openai_api_key="<HF_API_TOKEN>",
    openai_api_base="<ENDPOINT_URL>" + "/v1/",
)
llm.invoke("Why is open-source software important?")
```

> We’re able to directly leverage the same`ChatOpenAI` class that we would have used with the OpenAI models. This allows all previous code to function with our endpoint by changing just one line of code. 

Reviewer 3: ```suggestion
We’re able to directly leverage the same `ChatOpenAI` class that we would have used with the OpenAI models. This allows all previous code to work with our endpoint by changing just one line of code. 
```

> Let’s now use this declared LLM in a simple RAG pipeline to answer a question over the contents of a HF blog post.

Reviewer 3: ```suggestion
Let’s now use the LLM declared this way in a simple RAG pipeline to answer a question over the contents of a HF blog post.
```

)
from llama_index import download_loader
from llama_index.embeddings import HuggingFaceEmbedding
from llama_index.query_engine import CitationQueryEngine

SimpleWebPageReader = download_loader("SimpleWebPageReader")

documents = SimpleWebPageReader(html_to_text=True).load_data(
    ["https://huggingface.co/blog/open-source-llms-as-agents"]
)

# Load embedding model
embed_model = HuggingFaceEmbedding(model_name="BAAI/bge-large-en-v1.5")

# Pass LLM to pipeline
service_context = ServiceContext.from_defaults(embed_model=embed_model, llm=llm)
index = VectorStoreIndex.from_documents(
    documents, service_context=service_context, show_progress=True
)

# Query the index
query_engine = CitationQueryEngine.from_args(
    index,
    similarity_top_k=2,
)
response = query_engine.query(
    "According to this article which open-source model is the best for an agent behaviour?"
)
```

```
According to the article, Mixtral-8x7B is the best performing open-source model for an agent behavior [5]. It even beats GPT-3.5 in this task. However, it's worth noting that Mixtral's performance could be further improved with proper fine-tuning for function calling and task planning skills [5].
```

> ## Clean up

Reviewer 3: ```suggestion
## Cleaning up
```

> To clean up our work, we can either pause or delete the model endpoint. This step can alternately be completed via the UI. 

Reviewer 3: ```suggestion
After you are done with your endpoint, you can either pause or delete it. This step can be completed via the UI, or programmatically like follows. 
```
